### PR TITLE
2main/jmcguig

### PR DESCRIPTION
**Pull request by jmcguigan10**

### CODE REVIEW

File: model_training/ml_read_data.py [hunk 1]
```diff
@@ -23,19 +23,27 @@ def read_asymptotic_data(parms):
 
     dataset_train_list = []
     dataset_test_list = []
+    rng = torch.Generator().manual_seed(parms.get("random_seed", 0))
```
Comment (jmcguigan10) [inline model_training/ml_read_data.py]: given the same seed and same dataset length, code picks the same indices every time.

File: model_training/ml_read_data.py [hunk 2]
```diff
@@ -23,19 +23,27 @@ def read_asymptotic_data(parms):
...
+            growthrate = torch.Tensor(f_in["growthRate(1|s)"  ][...])
+            assert(parms["NF"] == int(np.array(f_in["nf"])) )
+
+        finite_mask = torch.isfinite(F4_initial).all(dim=(1,2,3)) & torch.isfinite(F4_final).all(dim=(1,2,3)) & torch.isfinite(growthrate)
+        if torch.any(~finite_mask):
+            n_bad = (~finite_mask).sum().item()
+            print("#   removing", n_bad, "non-finite samples")
+            F4_initial = F4_initial[finite_mask]
+            F4_final   = F4_final[finite_mask]
+            growthrate = growthrate[finite_mask]
```
Comment (jmcguigan10) [inline model_training/ml_read_data.py]: ### Switched HDF5 reads to 
```python
with h5py.File(...): 
```
so files are reliably closed on exceptions. Added ``` isfinite ``` masking to drop samples containing ```NaN/Inf``` in ```F4_initial, F4_final, or growthrate``` to prevent NaN loss/gradients. Assumes dim 0 is sample index and growthrate is aligned properly.

File: model_training/ml_read_data.py [hunk 3]
```diff
@@ -23,19 +23,27 @@ def read_asymptotic_data(parms):
...
 
         # downsample to less data
         if parms["samples_per_database"]>0:
             print("#    Downsampling to",parms["samples_per_database"],"samples")
-            random_indices = torch.randperm(len(growthrate))[:parms["samples_per_database"]]
+            random_indices = torch.randperm(len(growthrate), generator=rng)[:parms["samples_per_database"]]
```
Comment (jmcguigan10) [inline model_training/ml_read_data.py:41]: ### Passing ```generator=rng``` into
```python 
torch.randperm 
```
makes the downsampling deterministic given ```parms["random_seed"]```, instead of depending on whatever the global RNG state happened to be. Just making sure we're pinning proper processes.

File: model_training/ml_read_data.py [hunk 4]
```diff
@@ -90,14 +98,19 @@ def read_stable_data(parms):
...
     
+    rng = torch.Generator().manual_seed(parms.get("random_seed", 0))
     for filename in parms["stable_database_list"]:
-        f_in = h5py.File(filename,"r")
-        F4 = torch.squeeze(torch.Tensor(f_in["F4_initial(1|ccm)"][...]))
-        stable = torch.squeeze(torch.Tensor(f_in["stable"][...]))
-        f_in.close()
+        with h5py.File(filename,"r") as f_in:
+            F4 = torch.squeeze(torch.Tensor(f_in["F4_initial(1|ccm)"][...]))
+            stable = torch.squeeze(torch.Tensor(f_in["stable"][...]))
```
Comment (jmcguigan10) [inline model_training/ml_read_data.py:100]: ### Switched 
```python
h5py.File(...); close() 
```
to a 
```python
with h5py.File(...) as f_in: block, 
```
so the HDF5 file is always closed regardless of whether an error occurs in the loading process. This is what I read in documentation.

File: model_training/ml_read_data.py [hunk 5]
```diff
@@ -90,14 +98,19 @@ def read_stable_data(parms):
...
 
         # remove invalid data points indicated by a nan in the electron neutrino density
-        assert(torch.all(F4==F4))
+        finite_mask = torch.isfinite(F4).all(dim=(1,2,3)) & torch.isfinite(stable)
+        if torch.any(~finite_mask):
+            n_bad = (~finite_mask).sum().item()
+            print("#   removing", n_bad, "non-finite samples")
+            F4 = F4[finite_mask]
+            stable = stable[finite_mask]
```
Comment (jmcguigan10) [inline model_training/ml_read_data.py]: ### Instead of 
```python
assert(torch.all(F4==F4)) 
```
 we detect NaN/Inf, log how many samples are bad, and drop them from both F4 and stable so training can still carry on. Makes sure to print obviously so we know if some points are bad.

File: model_training/ml_trainmodel.py [hunk 6]
```diff
@@ -19,46 +19,57 @@
 # create an empty dictionary that will eventually contain all of the loss metrics of an iteration
 loss_dict = {}
 
-def configure_loader(parms, dataset_train_list):
+def configure_loader(parms, dataset_train_list, dataset_test_list):
+    assert len(dataset_train_list) == len(dataset_test_list)
+    assert len(dataset_train_list) > 0
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py]: Updated ```configure_loader``` to accept both train and test dataset lists, and added checks (non-empty, matching lengths) so we fail fast if the dataset pairing is wrong.

File: model_training/ml_trainmodel.py [hunk 7]
```diff
@@ -19,46 +19,57 @@
...
-    for dataset in dataset_train_list:
-        nsamples = len(dataset)
-        weights.extend([1.0/nsamples] * nsamples)
-    weights = torch.tensor(weights, dtype=torch.double)
+    weights = torch.cat(
+        [
+            torch.full((len(dataset),), 1.0 / len(dataset), dtype=torch.double)
+            for dataset in dataset_train_list
+        ]
+    )
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py]: Changed ```weights``` from a Python loop/list into a better torch tensor using ```torch.full + torch.cat```, keeping per-dataset uniform weighting but without the list stuff.

**Review by jmcguigan10 — COMMENTED** [review]
First Pull request onto main. Haven't run the tests on Jenkins yet.

File: model_training/ml_trainmodel.py [hunk 8]
```diff
@@ -201,33 +201,34 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
...
+                                                                                                                 traintest, "direction", direction_loss_fn)
+                    total_loss = total_loss + torch.exp(-model.log_task_weights["growthrate"]) * contribute_loss((growthrate_pred/ntot_invsec), #torch.log
+                                                                                                                 (growthrate_true/ntot_invsec), #torch.log
+                                                                                                                 traintest, "growthrate", comparison_loss_fn)
+                    if parms["do_unphysical_check"]:
+                        total_loss = total_loss + torch.exp(-model.log_task_weights["unphysical"]) * contribute_loss(F4f_pred/ntotal(F4i)[:,None,None,None],
+                                                                                                                     None,
+                                                                                                                     traintest, "unphysical", unphysical_loss_fn)
             return total_loss
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:249]: ### Made sure 
```python
accumulate_asymptotic_loss
```
 is in 
```python
torch.no_grad() 
```
mode— since the asymptotic losses are used for monitoring/logging, not optimization. I think this should prevent autograd from constructing and retaining a computation graph across predict_all() and downstream loss terms. I think this could help reduce VRAM usage and improve eval throughput while avoiding accidental gradient noise in our metrics.

File: model_training/ml_tools.py [hunk 9]
```diff
@@ -111,10 +111,10 @@ def restrict_F4_to_physical(F4_final):
     return F4_final
 
 def get_ndens_fluxmag_fhat(F4):
-    ntot = ntotal(F4)
+    ntot = torch.clamp(ntotal(F4), min=torch.finfo(F4.dtype).tiny)
     ndens = F4[:, 3,:,:]/ntot[:,None,None]
     flux  = F4[:,:3,:,:]/ntot[:,None,None,None]
-    fluxmag = torch.sqrt(torch.sum(flux**2, dim=1))
+    fluxmag = torch.clamp(torch.sqrt(torch.sum(flux**2, dim=1)), min=torch.finfo(F4.dtype).tiny)
```
Comment (jmcguigan10) [inline model_training/ml_tools.py]: Gradient Clipping.

Comment (srichers) [reply model_training/ml_tools.py]: why does this need to happen?

File: model_training/ml_trainmodel.py (hunk 6 above)
Comment (srichers) [reply model_training/ml_trainmodel.py]: I think the extra complexity of another function argument might not be worth it in this case. The loader should have a single clearly defined purpose: to load data. I think it sends the wrong message if one of the function arguments is data that it is not loading.

File: model_training/ml_trainmodel.py (hunk 8 above)
Comment (srichers) [reply model_training/ml_trainmodel.py:249]: Could we put this in the location where the function is used to reduce the amount of indentation?

File: model_training/ml_trainmodel.py [hunk 10]
```diff
@@ -226,17 +238,17 @@ def accumulate_asymptotic_loss(dataset_list, traintest):
...
+                    _, _, stable_pred = model.predict_all(F4i)
 
-                print(torch.sum(torch.abs(stable_pred-stable_true)).item()/stable_pred.shape[0],"fractional difference in stable points")
+                    print(torch.sum(torch.abs(stable_pred-stable_true)).item()/stable_pred.shape[0],"fractional difference in stable points")
 
-                # accumulate stability loss using MSE instead of BCE because it is easier to interpret
-                total_loss = total_loss + torch.exp(-model.log_task_weights["stability"] ) * \
-                    contribute_loss(stable_pred, stable_true, traintest, "stability", comparison_loss_fn)
+                    total_loss = total_loss + torch.exp(-model.log_task_weights["stability"] ) * \
+                        contribute_loss(stable_pred, stable_true, traintest, "stability", comparison_loss_fn)
```
Comment (srichers) [inline model_training/ml_trainmodel.py]: Similarly, I think it would be better to have no_grad where the function is used.

File: model_training/ml_read_data.py [hunk 11]
```diff
@@ -90,14 +98,19 @@ def read_stable_data(parms):
     dataset_train_list = []
     dataset_test_list = []
     
+    rng = torch.Generator().manual_seed(parms.get("random_seed", 0))
```
Comment (srichers) [inline model_training/ml_read_data.py]: I think it would be better to not have a default value. This forces all of the parameter choices to be intentional and prevents accidentally doing unintended things.

**Review by srichers — APPROVED** [review]
I removed the nan removal so it fails very loudly (helps prevent running lots of calculations without realizing something is wrong - the input data should absolutely not have any nans). I also removed the clipping of data values and a couple of other simplifications.

**Comment by jmcguigan10** [comment]
Ok, sounds good. I think failing loudly is good, but sometimes I prefer printing, or even assert if it's possible for the error to occur and not force the code to crash. Also the reason I thought clipping would be smart is because growth rate is a less important thing we're trying to predict, so I thought that would help a little. Not a big issue— it was just an idea I had.
